server/repository/repository.go

```diff
@@ -473,3 +532,36 @@ func (s *Server) testRepo(ctx context.Context, repo *appsv1.Repository) error {
 	})
 	return err
 }
+
+func (s *Server) isRepoPermittedInProject(repo string, projName string) error {
+	proj, err := s.projLister.Get(projName)
+	if err != nil {
+		return err
+	}
+	if !proj.IsSourcePermitted(appsv1.ApplicationSource{RepoURL: repo}) {
+		return status.Errorf(codes.PermissionDenied, "repository '%s' not permitted in project '%s'", repo, projName)
+	}
+	return nil
+}
+
+// isSourceInHistory checks if the supplied application source is either our current application
+// source, or was something which we synced to previously.
+func isSourceInHistory(app *v1alpha1.Application, source *v1alpha1.ApplicationSource) bool {
+	if source == nil {
+		return true
+	}
```

Comment:
When can `source` be `nil`? Shouldn't we return `false` here?

I feel this case at least needs some explanation in the comments.

Reply:
Since this endpoint is purely for UI purposes, it's really up to the UI to do the right thing. However, even if the UI is not supplying it, then the API call is not doing anything particularly dangerous -- namely supplying arbitrary values from source. 

NOTE: te have a check for `source == nil` at the top-level handler, so this is actually dead code. But that just means we can decide to return false with no consequence. So I will modify it to choose the safer option and explain with comment. Thanks for feedback.

Reply:
Actually, since we have the check, I'll just dereference it since the check for source == nil is done at the top of the function.